src/esp/assets/GenericInstanceMeshData.cpp

```diff
@@ -94,33 +105,183 @@ Cr::Containers::Optional<InstancePlyData> parsePly(
+        Mn::Vector3ub meshColor = meshColors[vertIdx];
+        // Convert color to an int @ vertex
+        const uint32_t meshColorInt = colorAsInt(meshColor);
+
+        std::unordered_map<uint32_t, std::pair<int, int>>::const_iterator
+            ssdColorToIDAndRegionIter =
+                tmpColorMapToSSDidAndRegionIndex.find(meshColorInt);
+
+        // derive semantic ID and color and region/room ID for culling
+        int regionID = 0;
+        int semanticID = 0;
+        if (ssdColorToIDAndRegionIter !=
+            tmpColorMapToSSDidAndRegionIndex.end()) {
+          // color is found in ssd mapping, so is legal color
+          semanticID = ssdColorToIDAndRegionIter->second.first;
+          regionID = ssdColorToIDAndRegionIter->second.second;
+        } else {
+          // color is not found in ssd mapping, so not legal color
+          // check if we've assigned a semantic ID to this color before, and if
+          // not do so
+          auto nonSSDClrRes =
+              nonSSDVertColors.insert({meshColorInt, nonSSDObjID});
+          if (nonSSDClrRes.second) {
+            // inserted, so increment nonSSDObjID
+            ++nonSSDObjID;
+          }
+          // map holds that color's nonSSDObjID
+          semanticID = nonSSDClrRes.first->second;
+          regionID = maxRegion;
+        }
+        // semantic ID for vertex
+        data.objectIds[vertIdx] = semanticID;
+        // color for
+        tmpDestSSDColorMap[semanticID] = meshColor;
+        // partition Ids for each vertex, for multi-mesh construction.
+        data.partitionIds[vertIdx] = regionID;
+      }
+
+      // color map with first nonSSDObjID-1 elements in proper order to match
+      // provided semantic IDs (so that ID is IDX of semantic color in map) and
+      // any overflow colors uniquely map 1-to-1 to an unmapped semantic ID as
+      // their index.
+      Cr::Containers::Array<Mn::Vector3ub> tmpDestSSDColorVec{
+          Mn::NoInit, tmpDestSSDColorMap.size()};
+      for (const auto& elem : tmpDestSSDColorMap) {
+        tmpDestSSDColorVec[elem.first] = elem.second;
+      }
+
+      // colors need to be in the appropriate order
+      colorMapToUse.resize(tmpDestSSDColorVec.size());
+      Cr::Utility::copy(tmpDestSSDColorVec,
+                        Cr::Containers::arrayView(colorMapToUse));
+
+    } else {
+      // no remapping provided by SSD so just use what is synthesized
+      Cr::Containers::Array<Mn::Color3ub> colorsThatBecomeTheColorMap{
+          Mn::NoInit, meshColors.size()};
+      // copy the meshColors into colorsThatBecomeTheColorMap
+      Cr::Utility::copy(meshColors, colorsThatBecomeTheColorMap);
+
+      // removeDuplicatesInPlace returns array of unique ids for colors, and
+      std::pair<Cr::Containers::Array<Mn::UnsignedInt>, std::size_t> out =
+          Mn::MeshTools::removeDuplicatesInPlace(
+              Cr::Containers::arrayCast<2, char>(
+                  stridedArrayView(colorsThatBecomeTheColorMap)));
```

Comment:
I still have a feeling that populating a unique color array could be done directly in the above loop, unifying the handling... :)

Reply:
Sure, we could only conditionally build the `tmpColorMapToSSDidAndRegionIndex` if the ssd mapping exists.  If it is empty, it would make a map full of "non-ssd" colors.  

Using this code path for this case would be slower than what we are doing now, and less elegant than the solution using the magnum tools.